### PR TITLE
Convert NodeList to Array to loop

### LINKED ISSUE #179: LiveView JS does not support older browsers
### Environment

* Elixir version (elixir -v): Elixir 1.8.1 (compiled with Erlang/OTP 21)
* Phoenix version (mix deps): 1.4.3
* NodeJS version (node -v): v6.11.0
* NPM version (npm -v): 3.10.10
* Operating system: iOS 9.3.5 (client)

### Actual behaviour

There are some JS issues that prevent LiveView working on an old iPad mini, which cannot be updated to a newer browser.

The initial page renders, but updates are not delivered.

There are at least two issues preventing this:

1. `NodeList.prototype.forEach` is not defined. This is surmountable: you can shim this quite easily before LiveSocket is loaded. (`NodeList.prototype.forEach = Array.prototype.forEach`)

2. `Element.prototype.classList` is a read-only property. In several places `this.el.classList` is assigned to, instead of added to/removed from. This seems to work (at least in newer versions of Safari), but https://developer.mozilla.org/en-US/docs/Web/API/Element/classList suggests that it should be read-only. It appears to me that using `this.el.className` would work (and have the same behaviour in other browsers).

### Expected behaviour

Older browsers should be able to run LiveView sites.

### PR DESCRIPTION
**Pull request by snewcomer**
This may seem verbose, but there are a few issues with other ways.

1.  using babel to do the work relies on Array.from with no older browser support.  It requires is to bring in a polyfill, which bumps the pkg size.
2. Could use a forEach function with a callback and scope, but then passing around scope is a pain.

See perf testing here: https://jsperf.com/nodelist-iterate/1

Luckily `length` exists on the nodeList on older browsers.  Downside is we are creating a new array. Lmk what everyone thinks!

NodeList forEach support 

Browser | Version
-- | --
Chrome | < 51
Firefox | <50
IE | *
Opera | <38
Safari | <10

close #179 
old PR #181

### CODE REVIEW

**Comment by basbz** [comment]
Afaik Array.prototype.slice.call(NodeList) will return an Array on all browsers including IE

**Comment by schinckel** [comment]
That seems to have fixed the JS errors I was getting in an older browser: however, it means that no websocket data is sent through, even on newer browsers. However, this may be entirely something I have done when attempting to update the dependency.

**Comment by schinckel** [comment]
Yeah, it must be something else I did: reverting back to the old version behaves the same way.

**Comment by schinckel** [comment]
I think maybe there is still one place that doesn't use the `toArray` helper: I'm still seeing a JS error. I believe it's in `joinRootViews`.

**Comment by schinckel** [comment]
(My use case: an old iPad was served fine by just using `NodeList.prototype.forEach = Array.prototype.forEach`, FWIW, but I don't know how far back this goes).

**Comment by schinckel** [comment]
Hmm. It seems for some reason my compiled version of the script is not updating correctly. I can see that file has been updated...

**Comment by schinckel** [comment]
I cleared out some stuff, and attempted to rebuild - I needed to `npm install morphdom` to get webpack to build stuff correctly, and after that I see `SyntaxError: Unexpected keyword 'const'. Const declarations are not supported in strict mode` from a webpack compiled file when loading in the browser.

Perhaps it's just not worth targeting stuff this old.

Thanks for your help though.

**Comment by schinckel** [comment]
(I think I could work with just the `classList` fix being applied, since a shim of `forEach` worked in my browser).

**Comment by snewcomer** [comment]
So to sum up this thread

- slice.call would also work.  Thx!  In the perf test below, slice.call was the slower compared to the for loop in Chrome (but faster than in Firefox).  I'm fine with either!

https://jsperf.com/nodelist-iterate

- I don't think you are the only person to encounter this given older Safari, Edge and IE don't support forEach on a NodeList.

Browser | Version
-- | --
Chrome | < 51
Firefox | <50
IE | *
Opera | <38
Safari | <10

**Comment by schinckel** [comment]
I think there is a subtle issue with using `classList.add` instead of `classList = `: something is not clearing `phx-error`. I'm not sure I can identify further what leads to that just yet.

**Comment by basbz** [comment]
I reacted to this issue after reading a blog post about `live_view` and landing at the repo.
The context wasn't totally clear.
When I look a the PR it feels a lot of copying is needed just to loop over a NodeList.
Why not forget about the copy and _handroll_  a forEach 
```
function each(items, fn) {
  for(let i = 0, len = items.length; i < len; i++) {
    fn(items[i]);
  }
}
```

**Comment by snewcomer** [comment]
yep that will totally work!  Did that in one of my iterations of this PR.
  
But with a hand-rolled forEach, you have to either pass around the context (`bind(this)`) or use an arrow function with the callback.  If you don't use an arrow function, it will troll you.  So very minor concern and we could go with your recommendation!  But b/c we don't have any strict linting rules (like always use arrow func) and/or lots of tests to cover everything, I left it out for now.  Perhaps I can add a comment about this though.

**Comment by basbz** [comment]
ok, but keeping the context should be hard.
`fn.call(this, node)`... will keep the binding as is
If needed I'd love to help out more with the Js is there a roadmap or something I can look to see what needs picking up?

**Comment by basbz** [comment]
Well it's is hard if isn't defined.
But libraries like underscore, lodash have always made it part of their api, to be able to define the context
https://underscorejs.org/#each

**Comment by snewcomer** [comment]
I wish 😞, but it is not the correct `this` in the scope of this hand rolled forEach.  Unless we put it on each class.  

Regarding the roadmap, I imagine after the initial version will have some opportunities!  But for now I believe, just trying to keep it dead simple and easy to work with.

**Comment by basbz** [comment]
I whipped up a quick poc.
I've replaced the arrows with functions to show how the context would work
https://github.com/snewcomer/phoenix_live_view/compare/sn/for-loop-array...basbz:sn/for-loop-array
I don't see why this couldn't work, and as a bonus save a lot of copying

**Comment by snewcomer** [comment]
Nice!  Yeah that way will work as well.  Thanks!  Doing a quick perf test, they seem about the same in a variety of browsers when you try it a few times.  So still good for whatever reads clearer and is simplest to understand for everybody 👍 

https://jsperf.com/each-fn-call

File: assets/js/phoenix_live_view.js [hunk 1]
```diff
@@ -182,6 +182,15 @@ let recursiveMerge = (target, source) => {
   }
 }
 
+let toArray = function(nodeList) {
+  let array = []
```
Comment (daytonn) [inline assets/js/phoenix_live_view.js]: Is there a reason not to use `Array.slice.apply` here?

```suggestion
  return [].slice.apply(nodeList)
```

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**